### PR TITLE
Replace all MCollective agent calls with the agent lib.

### PR DESCRIPTION
**Pull request by dhollinger**
This is an initial feature change. The goal was to mirror existing MCollective functionality, but using the Agent library. As the project moves along, additional functionality will be added.

Special thanks to @ripienaar for help.

### CODE REVIEW

File: lib/helpers/deployments.rb [hunk 1]
```diff
@@ -30,12 +36,24 @@ def deploy(branch, deleted)
...
-    command = if settings.use_mcollective
-                "#{settings.command_prefix} mco r10k deploy_module #{module_name} #{settings.mco_arguments}"
-              else
-                "#{settings.command_prefix} r10k deploy module #{module_name}"
-              end
-    message = run_command(command)
+    # Currently requires puppet_webhook to be run as a non-root
+    # user with access to running the MCollective client.
+    if settings.use_mcollective
+      result = PuppetWebhook::Mcollective.new('r10k',
```
Comment (dhollinger) [inline lib/helpers/deployments.rb]: This should probably be turned into a helper method to reduce duplicated code.

File: lib/helpers/deployments.rb [hunk 2]
```diff
@@ -1,17 +1,23 @@
...
   def deploy(branch, deleted)
-    if settings.use_mco_ruby
-      result = mco(branch).first
+    # Currently requires puppet_webhook to be run as a non-root
+    # user with access to running the MCollective client.
+    if settings.use_mcollective
+      result = PuppetWebhook::Mcollective.new('r10k',
+                                              'deploy',
+                                              {
+                                                dtimeoute: settings.discovery_timeout,
```
Comment (binford2k) [inline lib/helpers/deployments.rb]: I think this is a typo

File: lib/helpers/deployments.rb [hunk 3]
```diff
@@ -1,17 +1,23 @@
...
 
       message = result.results[:statusmsg]
     else
-      command = if settings.use_mcollective
-                  "#{settings.command_prefix} mco r10k deploy #{branch} #{settings.mco_arguments}"
-                else
-                  # If you don't use mcollective then this hook needs to be running as r10k's user i.e. root
-                  "#{settings.command_prefix} r10k deploy environment #{branch} #{settings.r10k_deploy_arguments}"
-                end
+      command = "#{settings.command_prefix} r10k deploy environment #{branch} #{settings.r10k_deploy_arguments}"
```
Comment (binford2k) [inline lib/helpers/deployments.rb]: Are all these variables sanitized already? eg, what if `branch` was set to `; rm -rf /;`?

File: lib/helpers/deployments.rb [hunk 4]
```diff
@@ -30,12 +36,24 @@ def deploy(branch, deleted)
...
-                "#{settings.command_prefix} r10k deploy module #{module_name}"
-              end
-    message = run_command(command)
+    # Currently requires puppet_webhook to be run as a non-root
+    # user with access to running the MCollective client.
+    if settings.use_mcollective
+      result = PuppetWebhook::Mcollective.new('r10k',
+                                              'deploy',
+                                              {
+                                                dtimeoute: settings.discovery_timeout,
```
Comment (binford2k) [inline lib/helpers/deployments.rb]: cut&paste, or is it really spelled like that? (if so, maybe a comment that the spelling is intentional)

Comment (dhollinger) [reply lib/helpers/deployments.rb]: Good catch. I just spelled that wrong.

File: lib/helpers/deployments.rb (hunk 3 above)
Comment (dhollinger) [reply lib/helpers/deployments.rb]: We are not sanitizing that, no. I'm not sure if that config parameter will still be around in the future. I'm planning on (attempting) to replace r10k open3 calls with the r10k libs (if possible).

TL;DR - I want to remove `run_command` if at all possible.

File: lib/helpers/deployments.rb [hunk 5]
```diff
@@ -1,17 +1,23 @@
...
+    if settings.use_mcollective
+      result = PuppetWebhook::Mcollective.new('r10k',
+                                              'deploy',
+                                              {
+                                                dtimeout: settings.discovery_timeout,
+                                                timeout: settings.client_timeout
+                                              },
+                                              settings.client_cfg,
+                                              environment: branch).run.first
       raise result.results[:statusmsg] unless result.results[:statuscode].zero?
```
Comment (ripienaar) [inline lib/helpers/deployments.rb]: you should also check the RPC stats that the client return and check for nodes that did not respond.  You can get to that via the client.stats method.

It'll be an instance of this https://github.com/puppetlabs/marionette-collective/blob/master/lib/mcollective/rpc/stats.rb and there check `noresponsefrom`

Comment (dhollinger) [reply lib/helpers/deployments.rb]: Good point. I've added some code for that.

@ripienaar @binford2k - could one/both of you give one more pass over before I merge? I added some code to both the plugin and the deployments helper methods based on @ripienaar's comment.

File: lib/helpers/deployments.rb [hunk 6]
```diff
@@ -1,17 +1,26 @@
...
+                                               {
+                                                 dtimeout: settings.discovery_timeout,
+                                                 timeout: settings.client_timeout
+                                               },
+                                               settings.client_cfg,
+                                               environment: branch).run
+      results.each do |result|
+        raise result.results[:statusmsg] unless result.results[:statuscode].zero?
+      end
+      raise results.stats[:noresponsefrom] unless result.stats[:noresponsefrom].length.zero?
```
Comment (ripienaar) [inline lib/helpers/deployments.rb]: seems good, though raising an array is super weird :P

Comment (dhollinger) [reply lib/helpers/deployments.rb]: Each RPC Result object is stored as an array and I couldn't think of a better way (at the time) to validate that each result status code was 0.
